src/Gateways/PagSeguro/Charge.php

```diff
@@ -0,0 +1,71 @@
+<?php
+
+namespace BeerAndCodeTeam\PaymentGateway\Gateways\PagSeguro;
+
+class Charge
+{
+    private string $referenceId;
+    private string $description;
+    private Amount $amount;
+    private PaymentMethodInterface $paymentMethod;
+    private string $notificationUrls = 'https://yourserver.com/nas_ecommerce/277be731-3b7c-4dac-8c4e-4c3f4a1fdc46/';
+
+    public function __construct(string $paymentMethod)
+    {
+        $this->amount = new Amount();
+        $this->amount->setCurrency('BRL');
+        switch ($paymentMethod) {
+            case 'CreditCard':
+                $this->paymentMethod = new CreditCard();
+                break;
+            case 'BankSlip':
+                $this->paymentMethod = new BankSlip();
+                break;
+            default:
+                throw new \Exception("Error: payment method {$paymentMethod} not existis");
+                break;
+        }
```

Comment:
Essa parte do código poderia ser algo assim:

```php
$validator = Validator::make(
['method'=>$paymentMethod],
['method'=> 'in:CrediCard,BankSlip']
);

if($validator->fails()){
    throw new \Exception("Error: payment method {$paymentMethod} not existis");
}

$this->amount = new Amount();
$this->amount->setCurrency('BRL');
$this->amount->paymentMethod = new $paymentMethod;
```

Apesar da sugestão, nenhuma dessas estruturas parece estar ok, a cada metodo de pagamento, o construct terá que ser alterado. Acho que o certo seria chamar outra classe com um metodo que faz validacao (baseada em algum tipo de listagem dinamica das classes de metodo) e criacao da instancia correta.

Ou caso sejam apenas esses mesmo e não veem problemas em alterar, então talvez a sugestão pode ajudar. De qualquer forma a verificação do parametro **$paymentMethod** deve vir antes que qualquer definição, afinal, não faz sentido criar uma instancia de *Amount* se o metodo não existe e será externada uma *Exception*.

Reply:
Se não me engano o PagSeguro só tem esses 2 métodos: boleto e cartão

Reply:
Nesse caso mantenho a sugestão que fiz quando a validação do parametro **$paymentMethod** (mas se a maioria optar pela que já esta implementada, tudo bem tbm). 

Outro ponto que mantenho é que essa validação (ou o case) fiquem antes das definições de **$this->amount**.